File: packages/core/src/quality-profile.ts
/**
 * Quality Profile — 品質預設模板
 *
 * 將 profile 名稱（strict/standard/minimal/none）展開為完整的 QualityConfig。
 * 支援使用者以 Partial<QualityConfig> 自訂覆寫。
 * 純函式、無副作用。
 */

import type { QualityConfig, QualityProfileName, TaskPlan } from "./types.js";

/**
 * 預設的 Quality Profile 定義
 */
const QUALITY_PROFILES: Record<QualityProfileName, QualityConfig> = {
  strict: {
    verify: true,
    judge_policy: "always",
    max_retries: 2,
    max_retry_budget_usd: 2.0,
  },
  standard: {
    verify: true,
    judge_policy: "on_change",
    max_retries: 1,
    max_retry_budget_usd: 1.0,
  },
  minimal: {
    verify: true,
    judge_policy: "never",
    max_retries: 0,
    max_retry_budget_usd: 0,
  },
  none: {
    verify: false,
    judge_policy: "never",
    max_retries: 0,
    max_retry_budget_usd: 0,
  },
};

/**
 * 判斷是否為 QualityProfileName
 */
function isProfileName(value: unknown): value is QualityProfileName {
  return typeof value === "string" && value in QUALITY_PROFILES;
}

/**
 * 解析 TaskPlan 的 quality 欄位，展開為完整的 QualityConfig
 *
 * - 字串 → 對應 profile 展開
 * - 物件 → 以 `none` 為基底，合併使用者自訂值
 * - undefined → `none`（向後相容）
 *
 * @param plan - 任務計畫
 * @returns 展開後的 QualityConfig
 */
export function resolveQualityProfile(plan: TaskPlan): QualityConfig {
  const quality = plan.quality;

  // 未設定 → none（向後相容）
  if (quality === undefined) {
    return { ...QUALITY_PROFILES.none };
  }

  // 字串 → profile 展開
  if (isProfileName(quality)) {
    return { ...QUALITY_PROFILES[quality] };
  }

  // 物件 → 以 none 為基底合併
  const base = { ...QUALITY_PROFILES.none };
  return {
    verify: quality.verify ?? base.verify,
    lint_command: quality.lint_command ?? base.lint_command,
    type_check_command: quality.type_check_command ?? base.type_check_command,
    judge_policy: quality.judge_policy ?? base.judge_policy,
    max_retries: quality.max_retries ?? base.max_retries,
    max_retry_budget_usd: quality.max_retry_budget_usd ?? base.max_retry_budget_usd,
  };
}

/**
 * 檢查品質相關警告
 *
 * 當 quality.verify 為 true 但 task 缺少 verify_command 時，產生警告。
 *
 * @param plan - 任務計畫
 * @param qualityConfig - 展開後的品質設定
 * @returns 警告訊息列表
 */
export function checkQualityWarnings(
  plan: TaskPlan,
  qualityConfig: QualityConfig,
): string[] {
  const warnings: string[] = [];

  if (!qualityConfig.verify) {
    return warnings;
  }

  const globalVerify = plan.defaults?.verify_command;

  for (const task of plan.tasks) {
    if (!task.verify_command && !globalVerify) {
      warnings.push(
        `Task ${task.id} 缺少 verify_command，但 quality profile 要求驗證。建議加上驗證指令或在 defaults 中設定。`,
      );
    }
  }

  return warnings;
}

File: python/devap/quality_profile.py
"""
Quality Profile — 品質預設模板

提供 strict / standard / minimal / none 四組預設品質設定。
"""

from __future__ import annotations

from typing import Optional

from devap.models.types import QualityConfig, QualityProfileName, TestPolicy

PROFILES: dict[QualityProfileName, QualityConfig] = {
    "strict": QualityConfig(
        verify=True,
        judge_policy="always",
        max_retries=3,
        max_retry_budget_usd=5.0,
    ),
    "standard": QualityConfig(
        verify=True,
        judge_policy="on_change",
        max_retries=2,
        max_retry_budget_usd=3.0,
    ),
    "minimal": QualityConfig(
        verify=True,
        judge_policy="never",
        max_retries=1,
        max_retry_budget_usd=1.0,
    ),
    "none": QualityConfig(
        verify=False,
        judge_policy="never",
        max_retries=0,
        max_retry_budget_usd=0.0,
    ),
}


def resolve_quality_profile(
    quality: QualityProfileName | QualityConfig | None,
    test_policy: Optional[TestPolicy] = None,
) -> QualityConfig:
    """
    解析品質設定

    將 profile 名稱展開為完整 QualityConfig，
    並合併 test_policy 的 static_analysis_command 和 completion_criteria。

    Args:
        quality: profile 名稱或自訂 QualityConfig
        test_policy: 測試策略（可選），會合併到 QualityConfig

    Returns:
        完整的 QualityConfig
    """
    if quality is None:
        config = PROFILES["none"].model_copy()
    elif isinstance(quality, QualityConfig):
        config = quality
    elif quality in PROFILES:
        config = PROFILES[quality].model_copy()
    else:
        config = PROFILES["none"].model_copy()

    # 合併 test_policy 到 QualityConfig（對齊 TS 端 quality-profile.ts）
    if test_policy is not None:
        if test_policy.static_analysis_command and not config.static_analysis_command:
            config.static_analysis_command = test_policy.static_analysis_command
        if test_policy.completion_criteria and not config.completion_criteria:
            config.completion_criteria = test_policy.completion_criteria

    return config
